Create schema

usingtechnology opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
# Description
Allow the user to create and publish a schema and an associated credential definition. 
This is not a PRODUCTION ready PR, it is to allow basic functionality. We will address this when we figure out where WE want it in the application.

Add in code/scripts to do full on local development and debugging.


<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

## Types of changes

<!-- What types of changes does your code introduce? Uncomment all that apply: -->

<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
Documentation (non-breaking change with enhancements to documentation)
<!-- Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [ ] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

Commit:
Add notes and scripts for local development (not just debugging into docker).

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
Need to correct the ledger browser configuration...

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
Need to correct how the dev agent connects to the dev/ide running bpa (which is not stood up by compose).

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
Add in ability to create a brand new schema on the ledger with a credential definition.

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
refactor createSchema method for code climate

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
use a better name for the param

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Jsyro commented on services/bb-pacman/backend/business-partner-agent/src/main/java/org/hyperledger/bpa/controller/AdminController.java:
```diff
@@ -84,6 +85,17 @@
+     * @param req {@link CreateSchemaRequest}
+     * @return {@link HttpResponse}
+     */
+    @Post("/schema/create")
```
Is there already a POST to '/schema'? It feels like this should be it.

parc-jason replied:
there is already one - it's for adding a schema (import).
this is for later refactoring when we figure out exactly what we want, just trying to not change much of their code at this time and introduce new bugs.

Jsyro commented on services/bb-pacman/backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/aries/SchemaService.java:
```diff
@@ -48,13 +53,61 @@
+            // for now hack this in, let's create a cred def with default values.
+            // no revocation!
+            CredentialDefinition.CredentialDefinitionRequest creddef = CredentialDefinition.CredentialDefinitionRequest.builder()
+                    .revocationRegistrySize(1000)
```
Should this be set if revocation is not supported?

parc-jason replied:
we are not supporting revocation, that requires the ledger have a tails server which isn't in place yet.

Jsyro replied:
line 96 defines the revocationRegistrySize... wonder if we should scrap that attribute too?

parc-jason replied:
changed to 4 because that is the minimum, cannot set to zero.

Jsyro commented on services/bb-pacman/scripts/local-development/.env-example:
```diff
@@ -0,0 +1,113 @@
+# ------------------------------------------------------------
```
Do you think these multi-service components should be inside the bb-pacman folder? I kinda figure they should be a higher level. If you like them here, let's delete the /docker/bbp-network folder at the repo root.

parc-jason replied:
nope. seems to me bbp-network is good for demos etc, these are for active development.

Commit:
For now, refresh your partner cred defs on startup and every 5 minutes.
We will need to refactor and get this done on demand.

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
Add in a little preventative care for creating new schemas.
Put in hooks to update the stored list of schemas/cred defs so we can add them to the wallet

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
Use events to rebuild the partner cred def list.
Add some logic to create schema to make sure it can be crypto signed.

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
fix path for production events
removed new package for tags, thought it was causing issues building docker image, but not the case... leaving out anyway as this is temporary code

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
using virtual env for ghg issuer, ignore the venv directory

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
issues building dockerfile using docker build , but somehow works with docker-compose, so have our composes output a local image to be reused when standing up multiple local instances.

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Commit:
refactor for code climate

Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Jsyro reviewed (dismissed):
Looks good to me, good work.
